ledger/participant-state-index/src/main/scala/com/daml/ledger/participant/state/index/impl/inmemory/InMemoryParticipantPartyRecordStore.scala

```diff
@@ -0,0 +1,203 @@
+    val attemptedUpdateF = withState(
+      state.get(party) match {
+        case Some(info) =>
+          for {
+            updatedInfo <- doUpdatePartyRecord(
+              partyRecordUpdate = partyRecordUpdate,
+              party = party,
+              info = info,
+            )
+          } yield toPartyRecord(updatedInfo)
+        case None =>
+          throw PartyRecordNotFoundOnUpdateException
+      }
+    ).map(tapSuccess { updatePartyRecord =>
+      logger.info(s"Updated party record in a participant local store: ${updatePartyRecord}")
+    })
+    attemptedUpdateF.recoverWith[Result[domain.ParticipantParty.PartyRecord]] {
+      case PartyRecordNotFoundOnUpdateException =>
+        for {
+          partyExistsOnLedger <- ledgerPartyExists.exists(party)
+          createdPartyRecord <-
+            if (partyExistsOnLedger) {
+              withState {
+                val newPartyRecord = domain.ParticipantParty.PartyRecord(
+                  party = party,
+                  metadata = domain.ObjectMeta.empty,
+                )
+                for {
+                  _ <- withoutPartyRecord(party = newPartyRecord.party) {
+                    doCreatePartyRecord(newPartyRecord)
+                  }
+                  _ <- withPartyRecord(party) { info =>
+                    doUpdatePartyRecord(
+                      partyRecordUpdate = partyRecordUpdate,
+                      party = party,
+                      info = info,
+                    ).map(toPartyRecord)
+                  }
+                  updatePartyRecord <- {
+                    withPartyRecord(party) { updatedInfo =>
+                      Right(toPartyRecord(updatedInfo))
+                    }
+                  }
+                } yield updatePartyRecord
+              }.map(tapSuccess { newPartyRecord =>
+                logger.error(
+                  s"Created a new party record in a participant local store: ${newPartyRecord}"
+                )
+              })(scala.concurrent.ExecutionContext.parasitic)
+            } else {
+              Future.successful(
+                Left(ParticipantPartyRecordStore.PartyNotFound(party))
+              )
+            }
+        } yield createdPartyRecord
+    }
+  }
+
+  private def doUpdatePartyRecord(
+      partyRecordUpdate: PartyRecordUpdate,
+      party: Party,
+      info: PartyRecordInfo,
+  ): Result[PartyRecordInfo] = {
+    val existingAnnotations = info.annotations
+    val updatedAnnotations =
+      partyRecordUpdate.metadataUpdate.annotationsUpdateO.fold(existingAnnotations) {
+        case AnnotationsUpdate.Merge(newAnnotations) => existingAnnotations.concat(newAnnotations)
+        case AnnotationsUpdate.Replace(newAnnotations) => newAnnotations
+      }
+    val currentResourceVersion = info.resourceVersion
+    val newResourceVersionEither = partyRecordUpdate.metadataUpdate.resourceVersionO match {
+      case None => Right(currentResourceVersion + 1)
```

Comment:
Right, I have re-read the API proposal that indeed has provisions for this field having been omitted.